core/src/smartcontracts/wasm.rs

```diff
@@ -285,11 +300,7 @@ impl<'wrld> Runtime<'wrld> {
         let alloc_fn = Self::get_alloc_fn(&mut caller)?;
         let memory = Self::get_memory(&mut caller)?;
 
-        // Accessing memory as a byte slice to avoid the use of unsafe
-        let query_mem_range = offset as usize..(offset + len) as usize;
-        let mut query_bytes = &memory.data(&caller)[query_mem_range];
-        let query =
-            QueryBox::decode(&mut query_bytes).map_err(|error| Trap::new(error.to_string()))?;
+        let query = Self::decode_from_memory(&memory, &mut caller, offset, len)?;
```

Comment:
```suggestion
        let query = Self::decode_from_memory(&mut caller, offset, len)?;
```

seems to me that you could do it this way

Reply:
also, can you use `encode_to_memory` when writing into memory inside this function?

Reply:
> seems to me that you could do it this way

To retrieve `memory` from `caller` we need to call `Self::get_memory()` which will try to find the exported memory. If I'm gonna to do that in `decode_from_memory()` I have to call `get_memory()`. And if some function wants to decode two objects from the same `memory` then there will be **2** calls for memory exporting instead of one. Exporting looks like a big job, so I decided to add a `memory` parameter.